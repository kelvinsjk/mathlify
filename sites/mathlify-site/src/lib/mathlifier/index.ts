/**
 * generates djot string,
 * using JavaScript's template literal syntax to handle mathematical expressions (both static and dynamic)
 *
 * math: starts with ${}, terminates with \n,
 * display: starts with $${}, terminates with \n\n,
 * display env: starts with $${'align'}, etc.
 * text: starts with @${}, terminates immediately
 */
export function mathlify(strings: TemplateStringsArray, ...values: unknown[]): string {
	return mathlifyGen(modules)(strings, ...values);
}

const modules: Modules = {
	math: (x: string) => `$\`${x}\``,
	display,
	mathEnvs: {
		equation: (x: string) => display(insertEnv('equation', x)),
		align: (x: string) => display(insertEnv('align', x)),
		gather: (x: string) => display(insertEnv('gather', x)),
		alignat: (x: string) => display(insertEnv('alignat', x)),
		equationStar: (x: string) => display(insertEnv('equation*', x)),
		alignStar: (x: string) => display(insertEnv('align*', x)),
		gatherStar: (x: string) => display(insertEnv('gather*', x)),
		alignatStar: (x: string) => display(insertEnv('alignat*', x)),
	},
};

function insertEnv(env: string, content: string, args = ''): string {
	if (env === 'alignat' || env === 'alignat*') {
		let firstLine = content.split('\n')[0];
		if (!firstLine) {
			firstLine = content.split('\n')[1];
		}
		const twoN = firstLine.split('&').length;
		if (twoN % 2 !== 0)
			console.warn(`alignat should have an even number of columns, but got ${twoN} instead`);
		args = `{${Math.floor(twoN / 2)}}`;
	}
	// todo: handle this with regex
	while (content.endsWith('\n') || content.endsWith('\t') || content.endsWith(' ')) {
		content = content.slice(0, content.length - 1);
	}
	return `\\begin{${env}}${args}${content}\n\\end{${env}}`;
}

function display(x: string): string {
	return `$$\`${x}\``;
}

enum Modes {
	text, // 0
	math, // 1
	display, // 2
	mathEnv, // 3
}

export interface Modules {
	math: (x: string) => string;
	display: (x: string) => string;
	mathEnvs?: {
		equation?: (x: string) => string;
		align?: (x: string) => string;
		gather?: (x: string) => string;
		alignat?: (x: string) => string;
		equationStar?: (x: string) => string;
		alignStar?: (x: string) => string;
		gatherStar?: (x: string) => string;
		alignatStar?: (x: string) => string;
	};
}

enum MathEnvs {
	equation,
	equationStar,
	align,
	alignStar,
	gather,
	gatherStar,
	alignat,
	alignatStar,
}

const mathEnvsKeys = [
	'equation',
	'equation*',
	'align',
	'align*',
	'gather',
	'gather*',
	'alignat',
	'alignat*',
] as const;
function isMathEnv(x: string): x is (typeof mathEnvsKeys)[number] {
	return (mathEnvsKeys as readonly string[]).includes(x);
}

const envToMode: Record<(typeof mathEnvsKeys)[number], MathEnvs> = {
	equation: MathEnvs.equation,
	'equation*': MathEnvs.equationStar,
	align: MathEnvs.align,
	'align*': MathEnvs.alignStar,
	gather: MathEnvs.gather,
	'gather*': MathEnvs.gatherStar,
	alignat: MathEnvs.alignat, // triggered by $${'alignat'}{x}
	'alignat*': MathEnvs.alignatStar, // triggered by $${'alignat'}{x}
};

interface MathEnvOptions {
	mathEnv: MathEnvs;
	eqnColumns?: number;
}

// TODO: alignat cols

export function mathlifyGen(
	modules: Modules,
): (strings: TemplateStringsArray, ...values: unknown[]) => string {
	return (strings: TemplateStringsArray, ...values: unknown[]) => {
		let currentModeString = '';
		let output = '';
		let mode = Modes.text;
		let mathEnvOptions: MathEnvOptions = { mathEnv: MathEnvs.equation };
		strings.forEach((str, i) => {
			const nextVal = values[i] ?? '';
			if (mode === Modes.text) {
				if (str.endsWith('@')) {
					// continues text mode
					currentModeString += `${str.slice(0, str.length - 1)}${nextVal}`;
				} else {
					// starts new text environment
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						str,
						'',
						nextVal,
						currentModeString,
					);
				}
			} else if (mode === Modes.math) {
				// checks for \n or \r\n non-greedily
				const regex = /([^]*?)(\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end math mode
					const [, before, newline, after] = match;
					output += `${before}`;
					if (output) {
						currentModeString += modules.math(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue math mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.math(output);
					}
				}
			} else if (mode === Modes.display) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end display mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += modules.display(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue display mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.display(output);
					}
				}
			} else if (mode === Modes.mathEnv) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end mathEnv mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue mathEnv mode
					if (str.endsWith(`@`)) {
						// TODO: check this.
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
				}
			}
		});
		return currentModeString;
	};
}

// returns [Mode, acc, curr, {mathEnv, eqnColumns?}]
function startNewEnv(
	after: string,
	newline: string,
	nextVal: unknown,
	curr: string,
): [Modes, string, string, MathEnvOptions] {
	const defaultMathEnvOptions = { mathEnv: MathEnvs.equation };
	let mathEnvOptions: MathEnvOptions = defaultMathEnvOptions;
	if (after.endsWith('$')) {
		let x = `${nextVal}`;
		let mode = Modes.display;
		if (isMathEnv(x)) {
			// new math env
			mode = Modes.mathEnv;
			mathEnvOptions = { mathEnv: envToMode[x] };
			//TODO: handle alignat cols
			x = '';
		}
		// new display mode
		return [mode, x, curr + `${newline}${after.slice(0, after.length - 1)}`, mathEnvOptions];
	} else if (after.endsWith('@')) {
		// new text mode
		return [
			Modes.text,
			'',
			curr + `${newline}${after.slice(0, after.length - 1)}${nextVal}`,
			defaultMathEnvOptions,
		];
	} else {
		// new math mode
		return [Modes.math, `${nextVal}`, curr + newline + after, defaultMathEnvOptions];
	}
}

function insertMathEnv(mathEnvOptions: MathEnvOptions, acc: string, modules: Modules): string {
	const mathEnv = mathEnvOptions.mathEnv;
	if (mathEnv === MathEnvs.equation) {
		return modules.mathEnvs?.equation ? modules.mathEnvs.equation(acc) : acc;
	} else if (mathEnv === MathEnvs.equationStar) {
		return modules.mathEnvs?.equationStar ? modules.mathEnvs.equationStar(acc) : acc;
	} else if (mathEnv === MathEnvs.align) {
		return modules.mathEnvs?.align ? modules.mathEnvs.align(acc) : acc;
	} else if (mathEnv === MathEnvs.alignStar) {
		return modules.mathEnvs?.alignStar ? modules.mathEnvs.alignStar(acc) : acc;
	} else if (mathEnv === MathEnvs.gather) {
		return modules.mathEnvs?.gather ? modules.mathEnvs.gather(acc) : acc;
	} else if (mathEnv === MathEnvs.gatherStar) {
		return modules.mathEnvs?.gatherStar ? modules.mathEnvs.gatherStar(acc) : acc;
	} else if (mathEnv === MathEnvs.alignat) {
		return modules.mathEnvs?.alignat ? modules.mathEnvs.alignat(acc) : acc;
	} else if (mathEnv === MathEnvs.alignatStar) {
		return modules.mathEnvs?.alignatStar ? modules.mathEnvs.alignatStar(acc) : acc;
	}
	console.warn(`mathEnv ${mathEnv} not found`);
	return acc;
}
