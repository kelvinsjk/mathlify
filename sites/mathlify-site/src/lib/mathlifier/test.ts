import { mathlify } from './';

const str = mathlify`This is inline math: ${'x+2'}xyz${'k'}
and this is text.

This is display math: $${'x+2'}xyz
abcd
${'k'}

This is aligned:
$${'align'}
  x+2 &= ${'k'} \\\\
  y &= ${'z'}`;

console.log(str);
