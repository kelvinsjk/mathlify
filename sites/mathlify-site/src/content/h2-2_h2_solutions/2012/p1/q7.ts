import { Answer } from '$lib/classes/answer';
import { QED } from '$lib/typesetting/utils';
import { mathlifierDj as mathlifier } from 'mathlifier';
import {
	generateAns,
	generateFracFnDefinition
} from '$content/h2-1_h2_learn/01_functions/04_more/01_self-inverse/02_practice';
import { expressionToPolynomial } from 'mathlify';
import svg from '$static/images/h2/fns/2012p1q7b.svg?raw';

export const answer = new Answer();

// f(x) = ax/(bx-a)
// g(x) = 1/x

const fState: Parameters<typeof generateFracFnDefinition>[0] = {
	// (bx+k)/(cx+a)
	fnType: 'frac',
	b: 1,
	c: 1,
	k: 'k',
	n: 2,
	show: true
};
const fExp = generateFracFnDefinition(fState, { for: true })[1];
// a
{
	const { soln } = generateAns(fState, fExp, 'g', { hideF2: true, qed: true });
	answer.addPart(mathlifier`[Show](#soln-1).`, soln);
}
// b
{
	const soln = `\`\`\` =html\n${svg}\n\`\`\`\n\n`;
	const ans = mathlifier`[Sketch](#soln-2).`;
	answer.addPart(ans, soln);
}
// c
{
	const [num, den] = fExp._getQuotientTerms();
	const { quotient, remainder, result } = expressionToPolynomial(num).longDivide(
		expressionToPolynomial(den)
	);
	const ans = mathlifier`Line of symmetry: ${{}} {y=x.}

Translate the curve ${{}} {y = \\frac{1}{x}}
by ${1}
unit in the positive ${{}} x\\text{-axis}
direction.
\\
Scale the resulting curve by a factor of ${remainder}
parallel to the ${{}} y\\text{-axis}.
\\
Translate the resulting curve by ${quotient}
unit in the positive ${{}} y\\text{-axis}
direction.`;
	const soln = mathlifier`Line of symmetry: ${{}} y = x \\; ${QED}

$${fExp} = ${result}

- Translate the curve ${{}} {y = \\frac{1}{x}}
by ${1}
unit in the positive ${{}} x\\text{-axis}
direction
- Scale the resulting curve by factor ${remainder}
parallel to the ${{}} y\\text{-axis}
- Translate the resulting curve by ${quotient}
unit in the positive ${{}} y\\text{-axis}
direction`;
	answer.addPart(ans, soln);
}
