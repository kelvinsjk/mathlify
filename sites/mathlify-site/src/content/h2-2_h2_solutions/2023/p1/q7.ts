import { Answer } from '$lib/classes/answer';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Polynomial } from 'mathlify';
import { absTerm } from 'mathlify/fns';
import { generateRange } from '$content/h2-1_h2_learn/01_functions/01_concepts/02_domain-and-range/02_practice-1';
import { absoluteRationalInverse } from '$content/h2-1_h2_learn/01_functions/02_inverse/03_formula/02_practice';
import { domainRestriction } from '$content/h2-1_h2_learn/01_functions/02_inverse/04_restriction/02_practice';
import { compositeExists } from '$content/h2-1_h2_learn/01_functions/03_composite/01_existence/02_practice';
import { QED } from '$lib/utils/typesetting/utils';
import svg from '$static/images/h2/fns/2023p1q7a.svg?raw';

export const answer = new Answer();

// ! Complete

// | (2x+4) / (3-x) |
// equivalent to | (bx+c) / (x+a) |
// from the learn/practice section

const [a, b, c] = [-3, 2, 4];
const num = new Polynomial([b, c]);
const den = new Polynomial([3, -1], { ascending: true });
const exp = absTerm([num, '/', den]);
const state: Parameters<typeof absoluteRationalInverse>[0] &
	Parameters<typeof domainRestriction>[0] = {
	fnType: 'abs',
	a,
	b,
	c,
	restriction: false,
	type: 'left',
	inclusive: true,
	zeroAns: true,
	zeroLeft: true,
	unknownConstants: false,
	definition: false
};
// from part (d) onwards
const restriction = { type: 'left', x: -2, inclusive: true } as const;

// a
{
	const soln = `\`\`\` =html\n${svg}\n\`\`\``;
	const ans = mathlifier`[Sketch](#soln-1).`;
	answer.addPart(ans, soln);
}
// b
const R_f = generateRange({ ...state, restriction: false }, exp).join(' \\cup ');
{
	const soln = mathlifier`$${{}} R_f = ${R_f} \\;${QED}`;
	const ans = mathlifier`${{}}R_f = ${R_f}.`;
	answer.addPart(ans, soln);
}
// c
{
	const { ans, soln } = compositeExists({ f: state, g: state, fg: true }, [exp, exp], {
		gName: 'f'
	});
	answer.addPart(ans, soln);
}
// d
{
	const { ans, soln } = domainRestriction(state, 'greatest');
	answer.addPart(ans, soln);
}
state.restriction = restriction;
// e
{
	const { ans, soln } = absoluteRationalInverse(state, [num, den], { swap: true });
	answer.addPart(ans, soln);
}
