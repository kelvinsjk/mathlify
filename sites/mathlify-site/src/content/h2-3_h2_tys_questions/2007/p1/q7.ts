/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { e, i, pi } from 'mathlify';
import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();

question.addBody(mathlifier`The polynomial ${'P(z)'}
has real coefficients. The equation ${{}} {P(z)=0}
has a root ${{}} r${e}^{${i}\\theta},
where ${{}} {r>0}
and ${{}} {0 < \\theta < ${pi}.}`);

// a
{
	question.addPart(
		mathlifier`Write down a second root in terms of ${'r'}
and ${'theta'},
and hence show that a quadratic factor of ${'P(z)'}
is $${{}} z^2 - 2rz \\cos \\theta + r^2.`,
		{ marks: 3 }
	);
}
// b
question.addPart(mathlifier`Out of syllabus (roots of unity).`, { marks: 4 });
// c
question.addPart(mathlifier`Out of syllabus (requires previous part).`, { marks: 3 });
