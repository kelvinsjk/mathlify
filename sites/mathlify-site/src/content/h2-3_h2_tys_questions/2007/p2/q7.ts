/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

const sumX = 4626;
const sumX2 = 147691;
const mu = 30;
const alpha = 5;
const n = 150;

export const question = new Question();

question.addBody(mathlifier`A large number of students in a
college have completed a geography project. The time, ${'x'}
hours, taken by a student to complete the project is noted
for a random sample of ${n}
students. The results are summarised by

$${{}} \\sum x = ${sumX}, \\quad \\sum x^2 = ${sumX2.toLocaleString()}.`);

// a
question.addPart(mathlifier`Find unbiased estimates of the population mean and variance.`, {
	marks: 2
});
// b
question.addPart(
	mathlifier`Test, at the ${alpha}\\%
significance level, whether the population mean time for a student to complete
the project exceeds ${mu}
hours.`,
	{ marks: 4 }
);
// c
question.addPart(
	mathlifier`State, giving a reason, whether any assumptions about
the population are needed in order for the test to be valid.`,
	{ marks: 1 }
);
