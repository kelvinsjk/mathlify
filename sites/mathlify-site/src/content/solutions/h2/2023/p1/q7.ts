import { Answer } from '$content/solutions/answerObject';
import { mathlify, mathlifyQED } from '$lib/mathlifier';

export const answer = new Answer();

// a
{
	// TODO: graphs
	const soln = mathlifyQED`Asymptotes: ${`y=2`}
and ${`x=3`}`;
	const ans = mathlify`Asymptotes: ${`y=2`}
and ${`x=3`}.`;
	answer.addPart(ans, soln);
}
