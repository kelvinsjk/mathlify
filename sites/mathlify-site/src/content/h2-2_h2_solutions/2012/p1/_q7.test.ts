import { expect, it } from 'vitest';
import { answer } from './q7';
import { mathlifierDj as mathlifier } from 'mathlifier';

const ans = answer.answer;
const actual: typeof ans = {
	parts: [
		{ body: mathlifier`[Show](#soln-1).` },
		{ body: mathlifier`[Sketch](#soln-2).` },
		{
			body: mathlifier`Line of symmetry: ${{}} {y=x.}

Translate the curve ${{}} {y = \\frac{1}{x}}
by ${1}
unit in the positive ${{}} x\\text{-axis}
direction.
\\
Scale the resulting curve by a factor of ${'1 + k'}
parallel to the ${{}} y\\text{-axis}.
\\
Translate the resulting curve by ${1}
unit in the positive ${{}} y\\text{-axis}
direction.`
		}
	]
};

it('2012p1q7', () => {
	expect(ans).toEqual(actual);
	expect(ans).toMatchSnapshot();
});
