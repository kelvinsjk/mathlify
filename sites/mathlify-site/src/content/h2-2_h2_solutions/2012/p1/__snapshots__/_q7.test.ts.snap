// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2012p1q7 1`] = `
{
  "parts": [
    {
      "body": "[Show](#soln-1).",
    },
    {
      "body": "[Sketch](#soln-2).",
    },
    {
      "body": "Line of symmetry: $\`  {y=x.} \`

Translate the curve $\`  {y = \\frac{1}{x}} \`
by $\` 1 \`
unit in the positive $\`  x\\text{-axis} \`
direction.
\\
Scale the resulting curve by a factor of $\` 1 + k \`
parallel to the $\`  y\\text{-axis}. \`
\\
Translate the resulting curve by $\` 1 \`
unit in the positive $\`  y\\text{-axis} \`
direction.",
    },
  ],
}
`;
